datameta/api/password.py

```diff
@@ -0,0 +1,88 @@
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+from pyramid.httpexceptions import HTTPNoContent, HTTPNotFound, HTTPForbidden, HTTPBadRequest
+
+from pyramid.view import view_config
+
+from .. import security
+from ..models import User, PasswordToken
+from uuid import UUID
+from datetime import datetime
+
+@view_config(
+    route_name="SetUserPassword",
+    renderer='json',
+    request_method="PUT",
+    openapi=True
+)
+def post(request):
+    db = request.dbsession
+
+    request_id = request.matchdict['id']
+    request_newPassword = request.openapi_validated.body["newPassword"]
+    request_credential = request.openapi_validated.body["passwordChangeCredential"]
+
+    # Case A: Password reset token
+    token = None
+    if request_id=='0':
+        token = db.query(PasswordToken).filter(PasswordToken.value==request_credential).one_or_none()
+        if token is not None and token.expires < datetime.now():
+            raise HTTPBadRequest(json=[{'message':'Token expired.'}])
+        auth_user = None if token is None else token.user
+        target_user = auth_user
+        if auth_user is None:
+            raise HTTPNotFound() # Token not found
+    # Case B: Authenticated user changing their own password
+    else:
+        # Obtain authorized user
+        auth_user = security.revalidate_user(request)
+
+        # Try to identify the target user by UUID
+        target_user = None
+        try:
+            target_user = db.query(User).filter(User.uuid==UUID(request_id)).one_or_none();
+        except ValueError:
+            pass
+        # Try to identify the target user by site ID
+        target_user = target_user if target_user is not None else db.query(User).filter(User.site_id==request_id).one_or_none();
+
+        if target_user is None:
+            raise HTTPNotFound() # User ID not found.
+
+        # Only changing the user's own password is allowed
+        if auth_user.id != target_user.id or not security.check_password_by_hash(request_oldPassword, auth_user.pwhash):
+            return HTTPForbidden() # Not authorized to change this user's password
+
+    # Verify the password quality
+    error = security.verify_password(request_newPassword)
+    if error:
+        return HTTPBadRequest(json_body=[{'message' : error}])
```

Comment:
Maybe rather HTTPForbidden? Maybe also custom error in `errors.py`

Reply:
`verify_password()` checks for password features, currently just length but maybe also things like "must contain at least one digit" etc. I didn't think that would be a use case for `403` but I may be wrong? 🤔

Reply:
Yes that's true, leave at 400. But could you then use the validation error from my PR?

Reply:
```
from ..errors import get_validation_error
...
raise get_validation_error(messages=["my message"])
```

Reply:
Done 👍